crates/bevy_math/src/sampling/statistical_tests/stats.rs

```diff
@@ -0,0 +1,320 @@
+        for (&index, &height) in self.inner.iter() {
+            let proj_index = [index[dimensions[0]], index[dimensions[1]]];
+            proj_hist
+                .entry(proj_index)
+                .and_modify(|v| *v += height)
+                .or_insert(height);
+        }
+
+        Ok(Histogram {
+            inner: proj_hist,
+            total: self.total,
+            invalid_count: self.invalid_count,
+        })
+    }
+}
+
+//------------//
+// Statistics //
+//------------//
+
+/// Compute the chi-squared goodness-of-fit test statistic for the `histogram` relative to the ideal
+/// distribution described by `ideal`. Note that this is distinct from the p-value, which must be
+/// assessed separately.
+pub fn chi_squared_fit(histogram: &Histogram<1>, ideal: &BinDistribution) -> f64 {
+    // The vector giving the number of hits expected in each bin for this many samples
+    let expecteds: Vec<_> = ideal
+        .bins
+        .iter()
+        .map(|p| (histogram.total as f64) * (*p as f64))
+        .collect();
+
+    let mut chi_squared: f64 = 0.0;
+    for (i, expected) in expecteds.into_iter().enumerate() {
+        let observed = histogram.get([i]) as f64;
+        let contribution = ((observed - expected) * (observed - expected)) / expected;
+        chi_squared += contribution;
+    }
+
+    chi_squared
+}
+
+/// Compute the chi-squared independence test statistic for the `histogram` relative to the ideal
+/// distributions described by `ideal`. Note that this is distinct from the p-value, which must be
+/// assessed separately.
+pub fn chi_squared_independence(histogram: &Histogram<2>, ideal: &[BinDistribution; 2]) -> f64 {
+    // Compute what the expected number of hits in each bin would be based on the total samples.
+    let mut expecteds: BTreeMap<[usize; 2], f64> = BTreeMap::new();
+    let [dist1, dist2] = ideal;
+    let (bins1, bins2) = (&dist1.bins, &dist2.bins);
+    let total_samples = histogram.total as f64;
+    for (i, &bin1) in bins1.iter().enumerate() {
+        for (j, &bin2) in bins2.iter().enumerate() {
+            expecteds.insert([i, j], bin1 as f64 * bin2 as f64 * total_samples);
+        }
+    }
+
+    let mut chi_squared: f64 = 0.0;
+    for (index, expected) in expecteds.into_iter() {
+        let observed = histogram.get(index) as f64;
+        let contribution = ((observed - expected) * (observed - expected)) / expected;
+        chi_squared += contribution;
+    }
+
+    chi_squared
+}
+
+//-----------------//
+// Critical Values //
+//-----------------//
+
+/// Critical values of the chi-squared distribution for ascending degrees of freedom at an alpha
+/// value of 0.001. This has been shifted by one so that the index corresponds to the degrees
+/// of freedom exactly; as a result, index zero is just a placeholder value.
+///
+/// Source: [NIST](https://www.itl.nist.gov/div898/handbook/eda/section3/eda3674.htm)
+pub const CHI_SQUARED_CRIT_VALUES_EMINUS3: [f64; 101] = [
```

Comment:
I'm a bit annoyed at the magic constants rather than computing these as needed, but I'm not sure that it's better.